synthesis/power_performance_area.proto

```diff
@@ -0,0 +1,165 @@
+syntax = "proto2";
+
+package hdl.ppa;
+
+// A word about units: The units used by the EDA-tools internally might vary
+// depending on context or as provided by the PDK.
+//
+// This proto is meant to always use standardized units so that it is
+// easy to compare; the program that is closest to the source of the data
+// reading logfiles and reports can convert these metrics once;
+// downstream we don't have to worry about it anymore.
+//
+// We're using
+//  - power units:  always watts (that reads easiest in scientific notation).
+//  - time units:   always picoseconds. This allows sufficient resolution to
+//                  store them as integer values.
+//  - length units: always micrometers (or micrometers squared for area).
+
+// A protocol buffer describing performance metrics for a particular design.
+message PowerPerformanceAreaProto {
+  optional Power power = 1;
+  optional Performance performance = 2;
+  optional Area area = 3;
+
+  // Information about the operating corner.
+  optional string cell_library = 4;
+  optional float operating_voltage = 5;
+
+  // Arbitrary string that might be used to describe the
+  // job that created this data, e.g. command line options.
+  optional string build_info = 10;
+}
+
+message Area {
+  // Length units here are abbreviated in ASCII-compatible
+  // way in the identifier suffixes.
+  //  `um`  - Micrometer (length)
+  //  `um2` - square Micrometer (area)
+
+  // The whole die area.
+  optional float die_area_um2 = 1;
+
+  // Area only covered by cells.
+  optional float cell_area_um2 = 2;
+
+  // If height and width are provided, this allows to
+  // calculate die_area_um2. Protocol buffer writers
+  // that know and provide width and height SHOULD also
+  // always provide die_area_um2.
+  // The following invariants hold:
+  //   die_area_um2  = die_height_um * die_width_um;
+  //   cell_area_um2 = die_area_um2 * cell_utilization_fraction
+  optional float die_height_um = 3;
+  optional float die_width_um = 4;
+
+  // A fraction of die_area_um2 used by cells. Range of [0..1]
+  optional float cell_utilization_fraction = 5;
+
+  // Leaving out proto tags 6..9 for later use.
+
+  // The total number of standard cells used.
+  // Invariance:
+  // total_cells = buffers + timing_buffers + flops + combinational;
+  optional int32 num_total_cells = 10;
+
+  // The number of combinational gate cells.
+  optional int32 num_combinational = 11;
+
+  // The number of non-timing buffers.
+  optional int32 num_buffers = 12;
+
+  // The number of timing buffers.
+  optional int32 num_timing_buffers = 13;
+
+  // The number of flops.
+  optional int32 num_flops = 14;
```

Comment:
I am reversing what I said earlier, perhaps "sequential" would be better instead of "flops". There are some sequential cells like synchronizers that are not flops, which may cause the total to add to total_cells. Sequential seems more generic, so we could be less precise so things still add up to the total.

Reply:
done.